### ISSUE TITLE
Connection and disconnection idempotence

### ISSUE THREAD

**Issue by sondrelg** [encode/databases#359]
Hi!

First of all thanks for creating and maintaining this! 👏   

### Question

The [connect](https://github.com/encode/databases/blob/master/databases/core.py#L82) and [disconnect](https://github.com/encode/databases/blob/master/databases/core.py#L105) core methods have an assertion each preventing repeat calls:

```python
assert not self.is_connected, "Already connected."
```

Would it make sense to replace these with an early return, to allow several repeat calls to the methods? Like this?

```python
if self.is_connected:
    return
```

### Motivation

Since I couldn't find any other issues or discussions for this I'm concerned I might be misusing the library in some way, but my root issue is basically this: 

I have an app with a few `databases` async context managers in different parts of my application - when testing this logic I sometimes want to create DB-fixtures which also requires me to open a connection. As soon as these mix, things break, which makes things complicated seems unnecessary (but I'm sure it might not be 🙂 )

-----------------

tl;dr: Would it be harmful to remove the asserts? It would make my life *a little* easier at least 😄

Id be happy to submit a PR if there aren't any reason why this shouldn't be allowed 👍

**Comment by aminalaee** [comment]
@sondrelg I see the point, but if you're still interested, can you provide some examples with this issue?
I don't see any harm in creating a PR.

**Comment by sondrelg** [comment]
In our test-setup we do something like this to create database resources (Starlette project), with pytest fixtures:

```python
# tests/conftest.py
@pytest.fixture
async def _db(_db_schema):
    if db.database.is_connected:
        await db.database.disconnect()

    if not db.database.is_connected:
        await db.database.connect()

    yield

    if db.database.is_connected:
        await db.database.disconnect()

@pytest.fixture
async def create_resource(_db):
    return await Resource.create(**{dict values})
```

And we also have this in our worker's main function:

```python
# worker.py // main function
    ...
    if not db.database.is_connected:
        await db.database.connect()
    
    try:
        # fire off a while loop
    except ...:
        ...
    finally:
        if db.database.is_connected:
            await db.database.disconnect()
```

_Without_ the library asserts I believe we could do this instead: 

```python
# tests/conftest.py
@pytest.fixture
async def _db(_db_schema):
    async with database.connection() as connection:
        yield

@pytest.fixture
async def create_resource(_db):
    return await Resource.create(**{dict values})
```
```python
# worker.py // main function    
    ...
    async with database.connection() as connection:
    
        try:
            # fire off a while loop
        except ...:
            ...
```

so tl;dr: it seems hard to avoid stacking connect statements when creating db fixtures, and it requires you to discard the nice context manager-style of connecting to the database; our code would be a little more elegant without the asserts in place. 

I guess the current behaviour does prevent accidental duplication of connect statements, but I'm not sure that's important - at least _I_ don't know of a reason why it would be 😛

Does that make sense @aminalaee?

**Comment by aminalaee** [comment]
@sondrelg That makes sense. I'm not sure how important that would be either. Maybe we can implement this and see if we're missing the point.
Are you interested to create a PR for that?

**Comment by sondrelg** [comment]
Yes I'd be happy to 👍I'll create one tonight 🥳
